Modules/socketmodule.c

```diff
@@ -876,7 +870,22 @@ init_sockobject(PySocketSockObject *s,
 {
     s->sock_fd = fd;
     s->sock_family = family;
+
+#ifdef __linux__
+    /* It's possible to pass SOCK_NONBLOCK and SOCK_CLOEXEC bit flags
+       on Linux as part of socket.type.  We want to reset them here,
+       to make socket.type be set to the same value on all platforms.
+       Otherwise, simple code like 'if sock.type == SOCK_STREAM' is
+       not portable.
+
+       0xF is SOCK_TYPE_MASK in include/linux/net.h which on Linux
+       is used to mask off SOCK_NONBLOCK and SOCK_CLOEXEC.
+    */
+    s->sock_type = type & 0xF;
```

Comment:
It's not possible to use SOCK_TYPE_MASK?

Reply:
No, Linux doesn't export it. It's a private macro in 'include/linux/net.h' that doesn't get exported in 'socket.h'.

Reply:
If it's not exported, maybe it's not a good idea to use it. Can we mask SOCK_NONBLOCK and SOCK_CLOEXEC instead? type & ~(SOCK_NONBLOCK | SOCK_CLOEXEC)

Reply:
Or maybe use getsockopt(SOL_SOCKET, SO_TYPE)? But this option would add a syscall.

Reply:
So `SOCK_TYPE_MASK` isn't exported.  I described why I think it's safe to just use `0xF` in the code in the issue, quoting:

> Since Linux doesn't export SOCK_TYPE_MASK I decided to hardcode it in socketmodule.c.  My reasoning:
>
> a/ It's highly unlikely that Linux will implement 5 more new socket types anytime soon, so 0xF should last for a *very* long time.
>
> b/ It's more likely than "a/" that they add a new flag like SOCK_CLOEXEC, in which case the "type & ~(SOCK_NONBLOCK | SOCK_CLOEXEC)" approach will fail.
>
> So given a/ and b/ I think it's safe to just use 0xF mask on Linux.

Reply:
> If it's not exported, maybe it's not a good idea to use it. Can we mask SOCK_NONBLOCK and SOCK_CLOEXEC instead? type & ~(SOCK_NONBLOCK | SOCK_CLOEXEC)

Alright, let's play it safe.  All in all, if Linux adds another 'SOCK_SOMETHING' bit flag, we'll add it to our `socket` module, and then we can add it to the list of excluded flags.

Reply:
At least, please don't hardcode a constant. Use something like "#define SOCK_TYPE_MASK 0xf". Maybe add a reference to bpo-32331.

In the module init, we may add the following assertion coming from the Linux kernel to make sure that SOCK_TYPE_MASK removes SOCK_CLOEXEC and SOCK_NONBLOCK flags:

```
    assert(!(SOCK_CLOEXEC & SOCK_TYPE_MASK));
    assert(!(SOCK_NONBLOCK & SOCK_TYPE_MASK));
```